### ISSUE TITLE
Image Classification Under Attack!

### ISSUE THREAD

**Issue by YueTang25** [YueTang25/deep-learning.github.io#2]
Authors: Yue Tang, Tianfang Fang

In the field of computer vision, image classification is a classic and popular research topic. From vanilla CNN to ResNet, the performance of deep-learning models on image classification has continuously improved, with state-of-the-art (SOTA) accuracy being enhanced multiple times.

However, despite the high accuracy of image classification, more and more people are beginning to focus on the lack of explainability in neural networks. How do these models work? Do they, like traditional machine learning models (such as tree-based models or regression-based models), learn to recognize specific features or attributes, or do they use other methods? For instance, when presented with an image of a frog, do neural networks learn that the frog is green, has smooth skin, and large eyes, or do they simply classify it as a "frog" in some high-dimensional space? This is an intriguing question.

Studying the weights of neural networks directly to answer this question can be time-consuming and difficult to interpret. So, we can take a different approach: we can "attack" the models. Specifically, by adding noise to images and comparing the classification scores of the original and noisy images, we can evaluate the robustness of the models.
For this analysis, we use the pretrained ResNet152 model from PyTorch as our image classifier and images from Unsplash to demonstrate the impact of adding noise.

To add noise, we apply a random number generated by np.random.normal to each pixel in the images and then output the processed images along with their classification scores. Below are some examples.

![image](https://github.com/user-attachments/assets/d3759bd1-aaa0-4dd4-9c2a-6b54e8e5877e)
 
From the output, we can observe that images with a single, clear main object and complete figures tend to be more resistant to noise. For example, the first image (a pug) maintains its classification with noise levels up to 0.43.
In contrast, the second image (either a Blenheim Spaniel or a Shih-Tzu), which only shows half of the head, is misclassified as a geyser when the noise level reaches 0.29.

For images that rely on context for correct classification—such as the example of a snorkel by a lakeside—performance drops significantly under attack. This is likely because these images depend heavily on environmental cues like water, and Gaussian noise can easily distort textures, making the model more vulnerable to noise.

Similarly, images that focus on fine details rather than full figures, like a hard disk, are also more easily fooled. Noise in such cases distorts or obscures critical details that the model uses for classification.

Another interesting observation is that as noise levels increase, the model tends to classify many noisy images as a fountain. This suggests that ResNet may have learned to associate certain textures or attributes of fountains with the appearance of Gaussian noise.

After this exploration, let's focus on another fascinating topic: generating adversarial examples for image classification. We can analyze the model's weights to generate images that are almost identical to the originals but are deliberately misclassified. This method was first proposed by Christian Szegedy and his co-authors in their 2014 paper _Intriguing Properties of Neural Networks_. Below are some examples from their work:

![image](https://github.com/user-attachments/assets/f7952925-e234-4175-af01-e47f80446023)

Interestingly, unlike random noise, adversarial examples look almost identical to the original images, yet they are mislabeled by the model. This is fascinating because it reveals that neural networks process information in images in a way that's quite different from humans—they seem to rely on indirect or abstract patterns.

We also applied the Carlini & Wagner (CW) attack to our own examples. Here are some of the resulting adversarial images:

<img src="https://github.com/user-attachments/assets/3c3c2834-b5b8-4625-ae42-6e41cffd901e" width="400" height="200"><img src="https://github.com/user-attachments/assets/61dbdc36-be31-459f-b378-fbc018d70a3b" width="400" height="200"><img src="https://github.com/user-attachments/assets/ce389195-3599-4b97-8230-dd27c33f8e99" width="400" height="200"><img src="https://github.com/user-attachments/assets/13ece015-9895-4cf4-bc25-81e4ad5b5189" width="400" height="200"><img src="https://github.com/user-attachments/assets/66edd95d-9bbf-4445-88e7-6e1d0e0b36ee" width="400" height="200">
While these examples may not be as effective as those in the paper, they still demonstrate how easily the model can be fooled.

The authors of the study also discovered that adding adversarial examples to the training dataset provides limited improvement. Despite this, we can still create adversarial examples. As they wrote in their conclusion: "_The existence of adversarial negatives appears to be in contradiction with the network’s ability to achieve high generalization performance. Indeed, if the network can generalize well, how can it be confused by these adversarial negatives, which are indistinguishable from the regular examples? A possible explanation is that the set of adversarial negatives is of extremely low probability, and thus is never (or rarely) observed in the test set, yet it is dense (much like the rational numbers), and so it is found near virtually every test case._"

From another perspective, we believe that adversarial examples may disrupt some inherent or natural distribution of images. This raises an interesting question: what if we used a diffusion model to process adversarial examples? Could this technique restore the images to their correct classifications? This is something we plan to explore in the future.
